wandb/jsonlfile.py

```diff
@@ -25,9 +25,41 @@ def column(self, key):
 
     def add(self, row):
         if not isinstance(row, collections.Mapping):
-            raise wandb.Error('history.add expects dict-like')
+            raise wandb.Error('history.add expects dict-like object')
         row['_runtime'] = time.time() - self._start_time
         self.rows.append(row)
         self._file.write(util.json_dumps_safer(row))
         self._file.write('\n')
         self._file.flush()
+
+
+class JsonlEventsFile(object):
+    """Used to store events during a run. """
+
+    def __init__(self, fname, out_dir='.'):
+        self._start_time = time.time()
+        self.fname = os.path.join(out_dir, fname)
+        self._file = open(self.fname, 'w')
+        self.buffer = []
+
+    def flatten(self, dictionary):
+        if type(dictionary) == dict:
+            for k, v in list(dictionary.items()):
+                if type(v) == dict:
+                    self.flatten(v)
+                    dictionary.pop(k)
+                    for k2, v2 in v.items():
+                        dictionary[k + "." + k2] = v2
+
+    def track(self, event, properties, timestamp=None, _wandb=False):
+        if not isinstance(properties, collections.Mapping):
```

Comment:
This function needs to be made thread-safe, right now it's not safe for users to use since our SystemStats thread while be using it. To make it thread safe just wrap in a lock